lib/topological_inventory/schema/default.rb

```diff
@@ -123,6 +123,63 @@ def add_tagging_collection(model, manager_ref: [:source_ref])
         end
       end
 
+      def add_service_instances
+        add_collection(:service_instances) do |builder|
+          add_default_properties(builder)
+          add_default_values(builder)
+
+          save_block = lambda do |source, inventory_collection|
+            src_refs = inventory_collection.data.collect { |inventory_object| inventory_object.source_ref }
+            return if src_refs.blank?
+
+            # Saving Service Instances as usual
+            InventoryRefresh::SaveCollection::Base.send(:save_inventory, inventory_collection)
+
+            # Get running tasks
+            tasks_source_ref = Task.where(:state => 'running', :target_type => 'ServiceInstance', :source_id => source.id)
+                                 .pluck(:target_source_ref)
+
+            # Load saved service instances (IDs needed)
+            svc_instances_values = ServiceInstance.where(:source_ref => tasks_source_ref).pluck(:id, :external_url, :source_ref, Arel.sql("extra->'finished'"), Arel.sql("extra->'status'"))
+            return if svc_instances_values.blank?
+
+            # Preparing Tasks data for mass update
+            sql_update_values = []
+            svc_instances_values.each do |attrs|
+              id, external_url, source_ref, finished_timestamp, status = attrs[0], attrs[1], attrs[2], attrs[3], attrs[4]
+
+              state = finished_timestamp.blank? ? 'running' : 'completed'
+              status = %w[error failed].include?(status) ? 'error' : 'ok' # TODO: ansible-tower specific, normalize in collector
+              context = {
+                :remote_status    => status,
+                :service_instance => {
+                  :id  => id,
+                  :url => external_url
+                }
+              }.to_json
+              sql_update_values << "('#{source_ref}', '#{state}', '#{status}', '#{context}'::json)"
+            end
+
+            sql = <<SQL
+UPDATE tasks AS t SET
+  state = c.state,
+  status = c.status,
+  context = c.context
+FROM (VALUES :values
+) AS c(source_ref, state, status, context)
+WHERE t.target_source_ref = c.source_ref
+  AND t.target_type = 'ServiceInstance'
+  AND t.state = 'running';
+SQL
```

Comment:
Yeah this isn't a good idea :)

Reply:
why not? raw sql update is much faster than using active record - and core is mostly about speed

Reply:
Because it is better to be easy to understand and maintain than to be fast 95% of the time.  If it turns out that this block of code is taking up a significant amount of time during a refresh we can optimize it then.  There is also a lot of optimizations that can be done within activerecord without dropping into pure sql.

Reply:
afaik there'll be similar upsert in Rails 6 (but not only update). Rails 5 doesn't have bulk update.

Reply:
@agrare is this refactoring correct for now as we discussed?

Reply:
I thought we agreed to go the active_record route and leave the equivalent sql commented out

Reply:
ok, I wasn't sure